### PR TITLE
Refactor VisibilityNotifier2D

### PR DESCRIPTION
**Pull request by reduz**
* Works from RenderingServer
* Accurately tells when on or off-screen, its no longer approximate.
* VisibilityEnabler2D also simplified to use the process mode instead.

**NOTE:** Only the 2D versions of the notifiers were refactored. Once this PR is merged, the 3D ones will be made to work the same.

### CODE REVIEW

File: scene/2d/visibility_notifier_2d.cpp [hunk 1]
```diff
@@ -127,235 +108,106 @@ void VisibilityNotifier2D::_bind_methods() {
 
 	ADD_PROPERTY(PropertyInfo(Variant::RECT2, "rect"), "set_rect", "get_rect");
 
-	ADD_SIGNAL(MethodInfo("viewport_entered", PropertyInfo(Variant::OBJECT, "viewport", PROPERTY_HINT_RESOURCE_TYPE, "Viewport")));
```
Comment (reduz) [inline scene/2d/visibility_notifier_2d.cpp:130]: These are gone, so its not possibleany longer to tell visibility per viewport. I dont know if anyone used this, but you will have to find a different workaround if this is the case. Its the cost to pay for a much simpler and more efficient system.

File: servers/rendering/renderer_canvas_cull.cpp [hunk 2]
```diff
@@ -173,32 +173,44 @@ void _attach_canvas_item_for_draw(RendererCanvasCull::Item *ci, RendererCanvasCu
 		}
 	}
 
-	if (ci->update_when_visible) {
-		RenderingServerDefault::redraw_request();
-	}
-
-	if ((ci->commands != nullptr && p_clip_rect.intersects(global_rect, true)) || ci->vp_render || ci->copy_back_buffer) {
+	if (((ci->commands != nullptr || ci->visibily_notifier) && p_clip_rect.intersects(global_rect, true)) || ci->vp_render || ci->copy_back_buffer) {
 		//something to draw?
```
Comment (lawnjelly) [inline servers/rendering/renderer_canvas_cull.cpp:177]: typo? Should be `visibility_notifier` presumably?

**Comment by lawnjelly** [comment]
Is this moving from a model of one `VisibilityNotifier` to many nodes, to one `VisibilityEnabler` to a single node, via `node_id`?

One thing I mentioned in #49381, is that as far as the user is concerned, having a separate node to do this is confusing, but does allow the `VisibilityNotifier` to be more efficient and operate on several nodes at once. If we remove the one to many relationship, could we instead put the functionality into an intermediate class that is derived from? That way we could maybe automatically activate it in some cases instead of having it so manually driven.

File: servers/rendering/renderer_canvas_cull.cpp (hunk 2 above)
Comment (reduz) [reply servers/rendering/renderer_canvas_cull.cpp:177]: ah I am so blind, thanks

**Comment by reduz** [comment]
@lawnjelly The new VisibilityEnabler works tightly with the new *process mode* in Godot 4. By default, nodes now have a complex way of specifying whether they want to process or not, and also by default they inherit the process mode from the parent node.

This is why, in this new version, just modifying the process mode of the parent node will affect all the children. If you want to change how some children are affected (like animations, physics, etc) you go and tweak the process mode of that node directly and it will automatically override what is set by VisibilityNotifier.

It just works different than in 3.x but, in practice, what you do is more or less the same.

**Review by lawnjelly — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by reduz** [comment]
Thanks for merging!

**Comment by Calinou** [comment]
For future reference, 2D audio with SubViewports was fixed by https://github.com/godotengine/godot/pull/76713.

**Pull request merged**